File: app/api/email_routes.py
from fastapi import APIRouter, HTTPException
from typing import List, Dict
from datetime import datetime

from app.services.nlp_service import analyze_email_thread
from app.services.gmail_service import fetch_threads
from app.services import nlp_service, priority_service
from app.services.thread_service import process_all_threads

from app.db.database import SessionLocal
from app.db.models import Email
from app.utils.pipeline_utils import (
    deduplicate_tasks,
    sort_by_priority,
    filter_low_confidence
)
import time
from concurrent.futures import ThreadPoolExecutor

CACHE = {
    "data": None,
    "timestamp": None
}

router = APIRouter()


# 🔹 SIMPLE ANALYZE (MANUAL INPUT)
@router.post("/analyze")
def analyze_email(data: dict):
    messages = data.get("messages", [])

    if not messages:
        raise HTTPException(status_code=400, detail="No messages provided")

    return analyze_email_thread(messages)


@router.get("/gmail")
def get_gmail_analysis(token: str):
    try:
        if not token or token == "null":
            raise HTTPException(status_code=400, detail="Invalid token")

        print("TOKEN RECEIVED:", token[:20])

        now = time.time()

        if token in CACHE and (now - CACHE[token]["timestamp"] < 60):
            print("⚡ Returning cached data")
            return CACHE[token]["data"]

        threads = fetch_threads(token, 20)

        def process_thread(t):
            messages = t.get("messages", [])
            thread_id = t.get("thread_id")

            if not messages:
                return None

            result = analyze_email_thread(messages)

            if not result or result.get("task") is None:
                return None

            result["thread_id"] = thread_id
            return result

        with ThreadPoolExecutor(max_workers=5) as executor:
            results = list(filter(None, executor.map(process_thread, threads)))

        results = filter_low_confidence(results)
        results = sort_by_priority(results)

        response = {
            "total_tasks": len(results),
            "tasks": results
        }

        CACHE[token] = {
            "data": response,
            "timestamp": now
        }

        return response

    except Exception as e:
        print("GMAIL ERROR:", e)
        raise HTTPException(status_code=500, detail=str(e))


# 🔥 THREAD PROCESSING API (ADVANCED)
@router.post("/process-all-threads")
def process_threads(emails: List[Dict]):
    try:
        results = process_all_threads(emails)

        for r in results:
            r["priority"] = priority_service.compute_priority(
                r.get("combined_text", ""),
                r.get("deadline"),
                "thread"
            )

        return results

    except Exception as e:
        print("THREAD ERROR:", e)
        raise HTTPException(status_code=500, detail=str(e))


# 🔥 SINGLE EMAIL + DB STORAGE
@router.post("/process-email")
def process_email(email: dict):
    db = SessionLocal()

    try:
        # convert timestamp safely
        timestamp = datetime.fromisoformat(email["timestamp"])

        new_email = Email(
            email_id=email["email_id"],
            thread_id=email["thread_id"],
            sender=email["sender"],
            subject=email["subject"],
            body=email["body"],
            timestamp=timestamp
        )

        db.add(new_email)
        db.commit()

        # NLP extraction
        result = nlp_service.extract_task_and_deadline(email["body"])

        # Priority calculation
        priority = priority_service.compute_priority(
            result["task"],
            result["deadline"],
            email["sender"]
        )

        return {
            "task": result["task"],
            "deadline": result["deadline"],
            "priority": priority
        }

    except Exception as e:
        db.rollback()
        print("EMAIL ERROR:", e)
        raise HTTPException(status_code=500, detail=str(e))

    finally:
        db.close()

File: backend/src/routes/email.routes.js
import { Router } from 'express';
import { analyzeEmailThread } from '../services/nlp.service.js';
import { fetchThreads } from '../services/gmail.service.js';
import { storeEmail } from '../services/vector.service.js';
import { filterLowConfidence, sortByPriority } from '../utils/pipeline.utils.js';
import { requireSession } from '../middleware/auth.middleware.js';

const router = Router();
const CACHE = new Map();
const CACHE_TTL = 60_000;

router.post('/analyze', requireSession, async (req, res) => {
  const { messages } = req.body;
  if (!messages?.length) {
    return res.status(400).json({ detail: 'No messages provided' });
  }
  try {
    res.json(await analyzeEmailThread(messages));
  } catch (err) {
    res.status(500).json({ detail: err.message });
  }
});

router.get('/gmail', requireSession, async (req, res) => {
  const { id: sessionId, accessToken, email: userEmail } = req.session;

  const now = Date.now();
  const cached = CACHE.get(sessionId);
  if (cached && now - cached.timestamp < CACHE_TTL) {
    return res.json(cached.data);
  }

  try {
    const threads = await fetchThreads(accessToken, 20);

    const settled = await Promise.allSettled(
      threads.map(async t => {
        if (!t.messages?.length) return null;
        const result = await analyzeEmailThread(t.messages);
        if (!result || result.task === null) return null;
        return { ...result, thread_id: t.thread_id };
      }),
    );

    let results = settled
      .filter(r => r.status === 'fulfilled' && r.value !== null)
      .map(r => r.value);

    results = filterLowConfidence(results);
    results = sortByPriority(results);

    const threadTextMap = Object.fromEntries(threads.map(t => [t.thread_id, t.messages[0]]));
    for (const r of results) {
      if (r.thread_id) {
        try {
          await storeEmail({
            user_email: userEmail,
            thread_id: r.thread_id,
            email_text: threadTextMap[r.thread_id] || '',
            task: r.task || '',
            deadline: r.deadline || '',
            priority: r.priority || 1,
            summary: r.summary || '',
            confidence: r.confidence || 0.5,
          });
        } catch (e) {
          console.warn('Vector store error:', e.message);
        }
      }
    }

    const response = { total_tasks: results.length, tasks: results };
    CACHE.set(sessionId, { data: response, timestamp: now });
    res.json(response);
  } catch (err) {
    console.error('GMAIL ERROR:', err.message);
    res.status(500).json({ detail: err.message });
  }
});

export default router;
